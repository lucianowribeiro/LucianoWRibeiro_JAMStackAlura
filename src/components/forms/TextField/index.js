/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import { get } from 'lodash';
import Text from '../../foundation/Text';

const InputWrapper = styled.div`
  margin-bottom: 17px;
`;

const Input = styled(Text)`
  width: 100%;
  color: ${({ theme, mode }) => get(theme, `${mode}.tertiary.main.color`)};
  background-color: ${({ theme, mode, color }) => (color
    ? get(theme, `${mode}.${color}.color`)
    : get(theme, `${mode}.background.main.color`))};
  border: 1px solid
    ${({ theme, mode }) => get(theme, `${mode}.tertiary.light.color`)};
  padding: 12px 16px;
  outline: 0;
  border-radius: ${({ theme }) => theme.borderRadius};
  ${({ theme, isFieldInvalid, mode }) => isFieldInvalid
    && css`
      border-color: ${get(theme, `${mode}.error.main.color`)};
      & + span {
        color: ${get(theme, `${mode}.error.main.color`)};
        font-size: 11px;
      }
    `}
`;

Input.defaultProps = {
  tag: 'input',
  variant: 'paragraph1',
};

export default function TextField({
  placeholder,
  name,
  onChange,
  value,
  error,
  isTouched,
  ...props
}) {
  const hasError = Boolean(error);
  const isFieldInvalid = hasError && isTouched;
  return (
    <InputWrapper>
      <Input
        type="text"
        placeholder={placeholder}
        name={name}
        onChange={onChange}
        value={value}
        isFieldInvalid={isFieldInvalid}
        {...props}
      />
      {isFieldInvalid && (
        <Text
          variant="smallestException"
          color="error.main"
          role="alert"
          {...props}
        >
          {error}
        </Text>
      )}
    </InputWrapper>
  );
}
TextField.defaultProps = {
  error: '',
  isTouched: false,
  value: undefined,
  onChange: undefined,
};

TextField.propTypes = {
  placeholder: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  error: PropTypes.string,
  isTouched: PropTypes.bool,
  onChange: PropTypes.func,
  value: PropTypes.string,
};
