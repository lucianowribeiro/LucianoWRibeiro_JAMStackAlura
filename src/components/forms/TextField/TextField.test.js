import React from 'react';
import { render, screen } from '../../../infra/test/testUtils';

import TextField from './index';

describe('<TextField />', () => {
  test('renders component', () => {
    render(
      <TextField
        placeholder="Nome"
        value="Luciano"
        onChange={() => {}}
        name="nome"
      />
      ,
    );
    // screen.debug();
    const textField = screen.getByPlaceholderText(/nome/i);
    expect(textField).toMatchSnapshot();
  });
});
