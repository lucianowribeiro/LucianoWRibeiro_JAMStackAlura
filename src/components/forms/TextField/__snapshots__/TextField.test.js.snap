// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders component 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
}

.c1 {
  width: 100%;
  border: 1px solid;
  padding: 12px 16px;
  outline: 0;
  border-radius: 12px;
}

<input
  class="c0 c1"
  color="tertiary.main"
  name="nome"
  placeholder="Nome"
  type="text"
  value="Ju"
/>
`;

exports[`<TextField /> when field is invalid displays the respective error message 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
}

.c1 {
  width: 100%;
  border: 1px solid;
  padding: 12px 16px;
  outline: 0;
  border-radius: 12px;
}

.c1 + span {
  font-size: 11px;
}

<input
  class="c0 c1"
  color="tertiary.main"
  name="email"
  placeholder="Email"
  type="text"
  value="devsoutinhogmail.com"
/>
`;
